Multiple Player improvements

alxbl opened the pull request:
- [X] Optional GUI dependencies (`pip install -U -e '.[GUI]'`)
- [X] Auto-play single replay files
- [x] Headless `pyrdp-player` (`--headless`)

Closes #151, Closes #163

obilodeau commented on bin/pyrdp-player.py:
```diff
@@ -63,19 +72,35 @@ def main():
-    mainWindow.show()
+    if not HAS_GUI and not args.headless:
+        logging.error('Headless mode is not specified and PySide2 is not installed. Install PySide2 to use the graphical user interface.')
+        exit(127)
```
The `exit` method exists in this context? Shouldn't this be `sys.exit()`?

alxbl replied:
It does, but I think it's an interpreter built-in, so it might not exist on all implementations of Python. I'll change it to `sys.exit`

obilodeau commented on pyrdp/player/HeadlessEventHandler.py:
```diff
@@ -0,0 +1,154 @@
+#
+# This file is part of the PyRDP project.
+# Copyright (C) 2019 GoSecure Inc.
```
```suggestion
# Copyright (C) 2020 GoSecure Inc.
```

alxbl commented:
> Question: I think the existing replay event handler didn't log file operations between the client and the server or mouse clicks (w/ coordinates) am I correct?

It doesn't. The mouse is rendered visually and it's pretty difficult/spammy to render the coordinates nicely in text only. We could try to log mouse clicks.

For the file events, we can add support in both places whenever we add it.

I'll make the modifications and merge this.

Cheers,

alxbl commented:
Oh lord. I just realized that the entire `keyboard.py` file which processes scancodes depends on Qt. I had reinstalled PySide2 so it didn't error, but my last tests are failing. I'm going to have to do a bit more work on this.

obilodeau commented:
> Oh lord. I just realized that the entire `keyboard.py` file which processes scancodes depends on Qt. I had reinstalled PySide2 so it didn't error, but my last tests are failing. I'm going to have to do a bit more work on this.

Great catch :+1: 

I was under the impression that we borrowed scan codes from FreeRDP but it could not be the case. Maybe if not using QT we could re-import them from FreeRDP. I let you investigate all the options.

obilodeau commented:
> It doesn't. The mouse is rendered visually and it's pretty difficult/spammy to render the coordinates nicely in text only. We could try to log mouse clicks.

Yes, I was thinking just on clicks you log the X, Y at that moment. Might seem not to add a lot of value but the way I see it is if I replay capture and have no event, I would be wondering if the user is doing everything through mouse whereas if I see the clicks then I would know for sure. This is out of scope for this PR btw.

> 
> For the file events, we can add support in both places whenever we add it.

:+1:

alxbl commented:
> I was under the impression that we borrowed scan codes from FreeRDP but it could not be the case. Maybe if not using QT we could re-import them from FreeRDP. I let you investigate all the options.

We have the scan code mappings to `Qt.Key` enums, I just need to make them not rely on the Qt import, so it should just be a matter of a few editor macros and minor code changes to use the new mappings.

alxbl commented:
Tested without PySide2 installed end-to-end to make sure I didn't miss anything. I fixed a few things in the process and added mouse click reporting in the replays (both headless and GUI).

I'll give you some time to look my latest commit over before merging.

I can see a few opportunities for refactoring (like having a better event handler hierarchy: `Headless -> Player -> Live`) and some code de-duplication, but I feel like this is clean enough right now.

robeving commented:
I'm excited to test this. I'm currently processing PCAPs of thousands of connections through some hacked up changed to pyrdp that use a framebuffered X server.

alxbl commented:
@robeving That sounds like a very creative workaround! Let me know how headless mode goes for you. Feel free to open issues if you run into any problems when trying it.

robeving commented:
It works and about 1000x faster!
